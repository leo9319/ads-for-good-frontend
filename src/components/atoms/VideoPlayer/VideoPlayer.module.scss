@import '../../../styles/breakpoints';

.videoContainer {
  width: inherit;
  height: inherit;

  .imageContainer {
    padding: 0;
    width: inherit;
    height: inherit;

    .buttonContainer {
      display: flex;
      justify-content: center;
      align-items: center;
      cursor: pointer;
      padding: 0;
      border: 0;
      background-color: transparent;
      width: inherit;
      height: inherit;

      &:disabled {
        cursor: default;
      }
    }

    .videoPlaceholder {
      position: relative;
      top: 0;
      left: 0;
      object-fit: cover;
      border-radius: 20px;
      color: #fff;
      background-color: #0f0f0f;
    }

    .playButton {
      position: absolute;
      width: 80px;
      height: 80px;
      z-index: 2000;

      @media (max-width: $lg-min) {
        width: 80px;
        height: 80px;
      }

      @media (max-width: $md-min) {
        width: 69px;
        height: 69px;
      }

      @media (max-width: $xs-min) {
        width: 44px;
        height: 44px;
      }

      &.sm {
        width: 44px;
        height: 44px;
      }

      &.md {
        width: 69px;
        height: 69px;
      }

      &.lg {
        width: 80px;
        height: 80px;
      }
    }
  }

  .embedContainer {
    padding: 0;
    width: inherit;
    height: inherit;

    iframe {
      position: relative;
      top: 0;
      left: 0;
      border-radius: 20px;
      border: none;
    }
  }

  .imageContainer .videoPlaceholder,
  .embedContainer iframe {
    max-width: 96vw;

    &.none {
      width: var(--videoWidth, 100vw);
      height: var(--videoHeight, 100vh);
    }

    &.lg {
      width: 1096px;
      height: 618px;
    }

    &.md {
      width: 944px;
      height: 532px;
    }

    &.sm {
      width: 352px;
      height: 198px;
    }
  }
}
