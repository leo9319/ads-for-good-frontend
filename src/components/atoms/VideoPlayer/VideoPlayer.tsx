import React, { ReactElement, useState } from 'react';

import Image from '@components/atoms/Image';
import Play from '@assets/icons/core/Play.svg';
import { classNames } from '@utils/common/classNames';
import FallbackImage from '@assets/images/FallbackImage.svg';

import styles from './VideoPlayer.module.scss';
import { getStylesAsCssProperties } from '@utils/common/styles';

const isYouTubeLink = (text: string): boolean => {
  const youtubeRegex =
    /(?:https?:\/\/)?(?:www\.)?(?:youtube\.com\/(?:watch\?v=|embed\/|v\/)|youtu\.be\/)([0-9A-Za-z_-]{11})/;
  return youtubeRegex.test(text);
};

const getSource = (videoId: string = '', attachment: string = '') =>
  `https://www.youtube.com/embed/${videoId}?${attachment}&autoplay=1&mute=1`;

const getVideoId = (url: string = ''): string => {
  const literal = url.lastIndexOf('v=') !== -1 ? 'v=' : '/';
  const videoId = (url.split(literal).pop() ?? '').split('?').shift() ?? '';
  return videoId ?? '';
};

const getYouTubeThumbnail = (videoId: string = ''): string => {
  const ampIndex = videoId.indexOf('&');
  if (ampIndex !== -1) {
    return `https://img.youtube.com/vi/${videoId.substring(0, ampIndex)}/hqdefault.jpg`;
  }
  return `https://img.youtube.com/vi/${videoId}/hqdefault.jpg`;
};

export interface VideoPlayerProps {
  /**
   * Source URL for the video, (mainly for youtube videos)
   * @property
   */
  src: string;
  /**
   * Size of the video container
   * @default 'none'
   * @property
   */
  size?: 'xl' | 'lg' | 'md' | 'sm' | 'none';
  /**
   * Width of the video
   * @property
   */
  width?: string;
  /**
   * Height of the video
   * @property
   */
  height?: string;
  /**
   * Thumbnail image for the video
   * @default ''
   * @property
   */
  thumbnail?: string;
  /**
   * Class name for the video thumbnail container
   * @default ''
   * @property
   */
  videoThumbnailContainerClassName?: string;
  /**
   * Class name for the video player container
   * @default ''
   * @property
   */
  videoPlayerContainerClassName?: string;
  /**
   * Class name for the video container
   * @default ''
   * @property
   */
  containerClassName?: string;
  /**
   * Class name for the video screen container
   * @default ''
   * @property
   */
  screenContainerClassName?: string;
  /**
   * Fallback image Source for the video, this is will work if actual video is broken
   * @default FallbackImage
   * @property
   */
  fallbackThumbnailSrc?: string;
}

/**
 * @group Atoms
 * @category Component
 *
 * VideoPlayer component is used to display a video with thumbnail and play button
 * @param {VideoPlayerProps} props
 * @returns {ReactElement}
 */
export const VideoPlayer = ({
  src,
  size = 'none',
  width,
  height,
  thumbnail = '',
  containerClassName,
  screenContainerClassName,
  videoPlayerContainerClassName,
  videoThumbnailContainerClassName,
  fallbackThumbnailSrc = FallbackImage,
}: VideoPlayerProps): ReactElement => {
  const [isVideoLoaded, setIsVideoLoaded] = useState(false);
  const adaptiveStyle =
    size === 'none'
      ? getStylesAsCssProperties({
          '--videoWidth': width,
          '--videoHeight': height,
        })
      : undefined;

  const videoId = getVideoId(src);
  const generatedThumbnail = () =>
    isYouTubeLink(src) && !isVideoLoaded ? getYouTubeThumbnail(videoId) : '';
  const videoThumbnail =
    thumbnail && thumbnail.trim() !== '' ? thumbnail : generatedThumbnail();
  const attachment = src?.split('?')?.pop() ?? '';
  const source = () => {
    if (!src) {
      return '';
    }
    if (isYouTubeLink(src)) {
      return getSource(videoId, attachment);
    }
    return src;
  };

  const videoSrc = source();
  const disbleVideo = !videoSrc;

  const handleLoadVideo = (
    event:
      | React.MouseEvent<HTMLButtonElement>
      | React.KeyboardEvent<HTMLButtonElement>
      | React.TouchEvent<HTMLButtonElement>
  ) => {
    if (!disbleVideo) {
      event.preventDefault();
      event.stopPropagation();
      setIsVideoLoaded(true);
    }
  };

  const onKeyDown = (event: React.KeyboardEvent<HTMLButtonElement>) => {
    const { key } = event;
    const keyPressed =
      !disbleVideo && key !== undefined && (key === 'Enter' || key === ' ');
    if (keyPressed) {
      handleLoadVideo(event);
    }
  };

  const onTouchEnd = (event: React.TouchEvent<HTMLButtonElement>) => {
    const { type } = event;
    const touched = !disbleVideo && type === 'touchend';
    if (touched) {
      handleLoadVideo(event);
    }
  };

  return (
    <div
      className={classNames(styles.videoContainer, containerClassName)}
      style={adaptiveStyle}
    >
      {isVideoLoaded ? (
        <div
          className={classNames(
            styles.embedContainer,
            videoPlayerContainerClassName
          )}
        >
          <iframe
            src={videoSrc}
            title="Video Player"
            className={classNames(styles[size], screenContainerClassName)}
            allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share; fullscreen;"
            referrerPolicy="strict-origin-when-cross-origin"
            allowFullScreen
          ></iframe>
        </div>
      ) : (
        <div
          className={classNames(
            styles.imageContainer,
            videoThumbnailContainerClassName
          )}
        >
          <button
            onKeyDown={onKeyDown}
            onTouchEnd={onTouchEnd}
            onClick={handleLoadVideo}
            disabled={disbleVideo}
            className={styles.buttonContainer}
          >
            <Image
              className={classNames(styles.playButton, 'button', styles[size])}
              alt="Play Button"
              src={Play}
              fallbackSrc={undefined}
            />
            <Image
              alt="Video Player"
              src={videoThumbnail}
              style={{ width, height }}
              className={classNames(
                styles.videoPlaceholder,
                styles[size],
                screenContainerClassName
              )}
              fallbackSrc={fallbackThumbnailSrc}
            />
          </button>
        </div>
      )}
    </div>
  );
};

export default VideoPlayer;
