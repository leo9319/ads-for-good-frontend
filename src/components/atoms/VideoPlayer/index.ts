export { VideoPlayer } from './VideoPlayer';
export { default } from './VideoPlayer';
export type { VideoPlayerProps } from './VideoPlayer';
