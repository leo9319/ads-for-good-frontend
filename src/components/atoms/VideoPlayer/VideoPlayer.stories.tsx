import type { Decorator, Meta, StoryObj } from '@storybook/react';
import VideoPlayer from '@components/atoms/VideoPlayer';
import React from 'react';
import { Theme } from '@radix-ui/themes';

const Decorators: Decorator = Story => (
  <Theme>
    <Story />
  </Theme>
);
// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: 'Library/Atoms/VideoPlayer',
  component: VideoPlayer,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: 'fullscreen',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    src: {
      description: 'Source URL for the video',
      control: { type: 'text' },
    },
    size: {
      control: {
        type: 'select',
      },
      options: ['lg', 'md', 'sm', 'none'],
    },
    width: {
      description: 'Width of the video',
      control: { type: 'text' },
    },
    height: {
      description: 'Height of the video',
      control: { type: 'text' },
    },
    thumbnail: {
      description: 'Thumbnail image for the video',
      control: { type: 'text' },
    },
    containerClassName: {
      description: 'Class name for the video container',
      control: { type: 'text' },
    },
    videoPlayerContainerClassName: {
      description: 'Class name for the video player container',
      control: { type: 'text' },
    },
    videoThumbnailContainerClassName: {
      description: 'Class name for the video thumbnail container',
      control: { type: 'text' },
    },
    screenContainerClassName: {
      control: { type: 'text' },
    },
    fallbackThumbnailSrc: {
      control: { type: 'text' },
    },
  },
  args: {
    size: 'none',
  },
  decorators: Decorators,
} satisfies Meta<typeof VideoPlayer>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
export const YoutubeVideo: Story = {
  args: {
    src: 'https://www.youtube.com/embed/J2ogs029GVg?si=GRnGAS3WvToc-CXL',
  },
};

export const VideoSourceStory: Story = {
  args: {
    src: 'https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4',
  },
};
