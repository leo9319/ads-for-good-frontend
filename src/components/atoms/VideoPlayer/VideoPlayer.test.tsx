import React from 'react';
import VideoPlayer from './VideoPlayer';
import { render } from '@testing-library/react';

describe('VideoPlayer', () => {
  it('should render successfully', () => {
    const { baseElement } = render(
      <VideoPlayer src="https://www.youtube.com/" />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
