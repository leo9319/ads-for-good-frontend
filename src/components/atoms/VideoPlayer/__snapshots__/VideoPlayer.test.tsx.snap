// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer should render successfully 1`] = `
<body>
  <div>
    <div
      class="videoContainer"
    >
      <div
        class="imageContainer"
      >
        <button
          class="buttonContainer"
        >
          <img
            alt="Play Button"
            aria-hidden="true"
            class="rt-Skeleton imgWrapper playButton button none"
            inert=""
            src="SvgrURL"
            tabindex="-1"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
